### PR TITLE
chore: allow to mark objects as immutable

### PR DESCRIPTION
**Pull request by triceo**
For Python solver, we need a way to mark certain objects as immutable.
We introduce a new internal interface for that purpose, and we make the cloner and the lookups honor the request.

### CODE REVIEW

File: core/src/main/java/ai/timefold/solver/core/impl/domain/lookup/LookUpStrategyResolver.java [hunk 1]
```diff
@@ -81,17 +38,17 @@ public LookUpStrategyResolver(DescriptorPolicy descriptorPolicy, LookUpStrategyT
      */
     public LookUpStrategy determineLookUpStrategy(Object object) {
         return decisionCache.computeIfAbsent(object.getClass(), objectClass -> {
-            if (objectClass.isEnum()) {
+            if (DeepCloningUtils.isImmutable(objectClass)) {
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/impl/domain/lookup/LookUpStrategyResolver.java:41]: I'm wondering... should we check for arrays here as well? Those can't possible be annotated.

Comment (Christopher-Chianelli) [reply core/src/main/java/ai/timefold/solver/core/impl/domain/lookup/LookUpStrategyResolver.java:41]: I  think value ranges that return arrays are rare, but probably not impossible. Arrays are mutable though, and may contain reference to mutable values (i.e. PlanningEntities).

**Review by Christopher-Chianelli — APPROVED** [review]
Changes LGTM

File: core/src/main/java/ai/timefold/solver/core/impl/domain/lookup/LookUpStrategyResolver.java (hunk 1 above)
Comment (triceo) [reply core/src/main/java/ai/timefold/solver/core/impl/domain/lookup/LookUpStrategyResolver.java:41]: That's true, but do they need to be looked up?
(The array check wouldn't be inside of `isImmutable`.)

Comment (Christopher-Chianelli) [reply core/src/main/java/ai/timefold/solver/core/impl/domain/lookup/LookUpStrategyResolver.java:41]: Yes, they will be looked up in `ChangeMove.rebase/SwapMove.rebase`.

**Pull request merged**